internal/otelmatchers/trace_matchers.go

```diff
@@ -0,0 +1,116 @@
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+func ConsistOfSpansWithIDs(expectedSpanIDs []pcommon.SpanID) types.GomegaMatcher {
+	return gomega.WithTransform(func(actual interface{}) ([]pcommon.SpanID, error) {
+		actualBytes, ok := actual.([]byte)
+		if !ok {
+			return nil, fmt.Errorf("ConsistOfSpansWithIDs rqquires a []byte, but got %T", actual)
+		}
+
+		actualTraces, err := unmarshalOTLPJSONTraces(actualBytes)
+		if err != nil {
+			return nil, fmt.Errorf("ConsistOfSpansWithIDs requires a valid OTLP JSON document: %v", err)
+		}
+
+		actualSpans := getAllSpans(actualTraces)
+
+		var actualSpanIDs []pcommon.SpanID
+		for _, span := range actualSpans {
+			actualSpanIDs = append(actualSpanIDs, span.SpanID())
+		}
+		return actualSpanIDs, nil
+	}, gomega.ConsistOf(expectedSpanIDs))
+}
+
+func EachHaveTraceID(expectedTraceID pcommon.TraceID) types.GomegaMatcher {
+	return gomega.WithTransform(func(actual interface{}) ([]pcommon.TraceID, error) {
+		actualBytes, ok := actual.([]byte)
+		if !ok {
+			return nil, fmt.Errorf("EachHaveTraceID rqquires a []byte, but got %T", actual)
+		}
+
+		actualTraces, err := unmarshalOTLPJSONTraces(actualBytes)
+		if err != nil {
+			return nil, fmt.Errorf("EachHaveTraceID requires a valid OTLP JSON document: %v", err)
+		}
+
+		actualSpans := getAllSpans(actualTraces)
+
+		var actualTraceIDs []pcommon.TraceID
+		for _, span := range actualSpans {
+			actualTraceIDs = append(actualTraceIDs, span.TraceID())
+		}
+		return actualTraceIDs, nil
+	}, gomega.HaveEach(expectedTraceID))
+}
+
+func EachHaveAttributes(expectedAttrs pcommon.Map) types.GomegaMatcher {
+	return gomega.WithTransform(func(actual interface{}) ([]map[string]any, error) {
+		actualBytes, ok := actual.([]byte)
+		if !ok {
+			return nil, fmt.Errorf("EachHaveAttributes rqquires a []byte, but got %T", actual)
+		}
+
+		actualTraces, err := unmarshalOTLPJSONTraces(actualBytes)
+		if err != nil {
+			return nil, fmt.Errorf("EachHaveAttributes requires a valid OTLP JSON document: %v", err)
+		}
+
+		actualSpans := getAllSpans(actualTraces)
+
+		var actualAttrs []map[string]any
+		for _, span := range actualSpans {
+			actualAttrs = append(actualAttrs, span.Attributes().AsRaw())
+		}
+		return actualAttrs, nil
+	}, gomega.HaveEach(gomega.Equal(expectedAttrs.AsRaw())))
+}
+
+func getAllSpans(traces []ptrace.Traces) []ptrace.Span {
+	var spans []ptrace.Span
+
+	for _, td := range traces {
+		for i := 0; i < td.ResourceSpans().Len(); i++ {
+			resourceSpans := td.ResourceSpans().At(i)
+			for j := 0; j < resourceSpans.ScopeSpans().Len(); j++ {
+				scopeSpans := resourceSpans.ScopeSpans().At(j)
+				for k := 0; k < scopeSpans.Spans().Len(); k++ {
+					spans = append(spans, scopeSpans.Spans().At(k))
+				}
+			}
+		}
+	}
+
+	return spans
+}
+
+func unmarshalOTLPJSONTraces(buf []byte) ([]ptrace.Traces, error) {
+	var results []ptrace.Traces
+
+	tracesUnmarshaler := &ptrace.JSONUnmarshaler{}
```

Comment:
I see no reason for using a reference here. Can be simplified down to:

```go
tracesUnmarshaler := ptrace.JSONUnmarshaler{}
```

Reply:
Agree